File: throwcode_python/algorithm/sort/bubble_sort.py
#!/use/bin/env/python3
# -*- coding:utf-8 -*-
# @author: goinhn
# @date: 2021/2/25 23:57

def bubble_sort(arr):
    """
    冒泡排序
    :param arr: 待排序的数组
    :return:
    """
    if arr is None or len(arr) < 2:
        return
    for i in range(len(arr) - 1):
        for j in range(len(arr) - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]


def main():
    arr = [2, 3, 1, 5, 4]
    bubble_sort(arr)
    for value in arr:
        print(value)


if __name__ == '__main__':
    main()

File: throwcode_java/src/main/java/com/goinhn/algorithm/sort/BubbleSort.java
package com.goinhn.algorithm.sort;

/**
 * <p>
 * 冒泡排序
 * 稳定排序
 * </p>
 * <p>
 * 时间复杂度 O(N^2)
 * 空间复杂度 O(1)
 * </p>
 *
 * @author goinhn
 * @date 2020-08-23T0:09:00
 */
public class BubbleSort {

    public static void main(String[] args) {
        int[] arr = new int[]{1, 2, 4, 1, 3, 5};
        bubbleSort(arr);
        for (int i : arr) {
            System.out.println(i);
        }
    }

    /**
     * 冒泡排序
     *
     * @param arr
     */
    public static void bubbleSort(int[] arr) {
        if (arr == null || arr.length < 2) {
            return;
        }
        for (int i = arr.length - 1; i > 0; i--) {
            for (int j = 0; j < i; j++) {
                if (arr[j] > arr[j + 1]) {
                    swap(arr, j, j + 1);
                }
            }
        }
    }

    /**
     * 交换方法
     *
     * @param arr
     * @param x
     * @param y
     */
    public static void swap(int[] arr, int x, int y) {
        int temp = arr[x];
        arr[x] = arr[y];
        arr[y] = temp;
    }

}
